### PR TITLE
fix(policy): Use search to fetch all policies

### PR DESCRIPTION
**Pull request by dexter-mh-lee**
Currently, we use listUrns function to fetch all policies. We realized this function does not scale well when a lot of entities have been ingested. 

Instead, we will start using search to fetch all policies. Caveat, we have no Searchable annotations on any fields for policies, which means that the search index is currently empty. Modified the ingestPoliciesStep to send MCLs for the existing policies if the search index is empty, so that we fill up the search index before doing any fetching. 

As a side, add a new searchable field called lastUpdatedTimestamp, and set it on any updates. List policies sorts based on this field so newly upserted policies get ranked above.

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-web-react/src/app/policy/PoliciesPage.tsx [hunk 1]
```diff
@@ -244,6 +237,9 @@ export const PoliciesPage = () => {
             content: `Are you sure you want to remove policy?`,
             onOk() {
                 deletePolicy({ variables: { urn: policy?.urn as string } }); // There must be a focus policy urn.
+                setTimeout(function () {
+                    policiesRefetch();
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/policy/PoliciesPage.tsx:241]: nice!

File: datahub-web-react/src/app/policy/PoliciesPage.tsx [hunk 2]
```diff
@@ -282,6 +278,9 @@ export const PoliciesPage = () => {
             createPolicy({ variables: { input: toPolicyInput(savePolicy) } });
         }
         message.success('Successfully saved policy.');
+        setTimeout(function () {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/policy/PoliciesPage.tsx:281]: nice!

File: metadata-models/src/main/pegasus/com/linkedin/policy/DataHubPolicyInfo.pdl [hunk 3]
```diff
@@ -48,4 +51,11 @@ record DataHubPolicyInfo {
    */
   editable: boolean = true
 
+  /**
+   * Timestamp when the policy was last updated
+   */
+  @Searchable = {
+    "fieldType": "DATETIME"
+  }
+  lastUpdatedTimestamp: optional long
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/policy/DataHubPolicyInfo.pdl:60]: Thoughts on lastModifiedMs -- only cause timestamp can mean a few formats, also to align with startTimeMs in the ExecutionRequestResult model that we use for sorting ingestion runs. Not a huge deal but let me know what you think

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java [hunk 4]
```diff
@@ -266,19 +249,8 @@ public void run() {
     }
 
     private void addPoliciesToCache(final Map<String, List<DataHubPolicyInfo>> cache,
-        final Collection<EntityResponse> entityResponses) {
-      for (final EntityResponse entityResponse : entityResponses) {
-        addPolicyToCache(cache, entityResponse);
-      }
-    }
-
-    private void addPolicyToCache(final Map<String, List<DataHubPolicyInfo>> cache, final EntityResponse entityResponse) {
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:275]: Nice! I love the idea of simplifying this.

File: metadata-models/src/main/pegasus/com/linkedin/policy/DataHubPolicyInfo.pdl (hunk 3 above)
Comment (dexter-mh-lee) [reply metadata-models/src/main/pegasus/com/linkedin/policy/DataHubPolicyInfo.pdl:60]: Oh I just followed the one added to Operation

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyFetcher.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    if (policyUrns.isEmpty()) {
+      return new PolicyFetchResult(Collections.emptyList(), 0);
+    }
+
+    // Fetch DataHubPolicyInfo aspects for each urn
+    final Map<Urn, EntityResponse> policyEntities =
+        _entityClient.batchGetV2(POLICY_ENTITY_NAME, new HashSet<>(policyUrns), null, authentication);
+    return new PolicyFetchResult(policyUrns.stream()
+        .map(policyEntities::get)
+        .filter(Objects::nonNull)
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyFetcher.java:57]: qq - do we need double non null filter?

Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyFetcher.java:57]: what if we just filtered for wheree apect map contains key

Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyFetcher.java:57]: (minor)

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java [hunk 6]
```diff
@@ -88,11 +98,45 @@ public void execute() throws IOException, URISyntaxException {
     log.info("Successfully ingested default access policies.");
   }
 
+  /**
+   * Send MCLs for each policy to refill the policy search index
+   */
+  private void sendMCL() throws URISyntaxException {
+    log.info("Pushing MCLs for all policies");
```
Comment (jjoyce0510) [inline metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java:105]: what policies? bootstrap policies? do they already exist inside the document store?

File: metadata-service/restli-client/src/main/java/com/linkedin/entity/client/EntityClient.java [hunk 7]
```diff
@@ -146,14 +146,15 @@ public ListResult list(@Nonnull String entity, @Nullable Map<String, String> req
...
+   * @param sortCriterion sort criterion
    * @param start start offset for search results
    * @param count max number of search results requested
    * @return Snapshot key
    * @throws RemoteInvocationException
    */
   @Nonnull
-  public SearchResult search(@Nonnull String entity, @Nonnull String input, @Nullable Filter filter, int start,
-      int count, @Nonnull Authentication authentication) throws RemoteInvocationException;
+  public SearchResult search(@Nonnull String entity, @Nonnull String input, @Nullable Filter filter,
```
Comment (jjoyce0510) [inline metadata-service/restli-client/src/main/java/com/linkedin/entity/client/EntityClient.java:156]: nice! We also have a filter method which i've used for the same purpose.

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyFetcher.java (hunk 5 above)
Comment (dexter-mh-lee) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyFetcher.java:57]: so just one filter that checks whether urns exist in the map and then entity response has the aspect? feel like efficiency wise this should be exactly the same

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java (hunk 6 above)
Comment (dexter-mh-lee) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java:105]: Leaving for record. The current policy index is empty bc there are no searchable fields. So this step makes sure we reingest the existing policies so they get in the index one time

**Review by jjoyce0510 — APPROVED** [review]
overall LGTM - let's test on demo extensively before we cut a release :p

File: metadata-models/src/main/pegasus/com/linkedin/policy/DataHubPolicyInfo.pdl (hunk 3 above)
Comment (dexter-mh-lee) [reply metadata-models/src/main/pegasus/com/linkedin/policy/DataHubPolicyInfo.pdl:60]: Found it https://github.com/datahub-project/datahub/blob/master/metadata-models/src/main/pegasus/com/linkedin/common/Operation.pdl#L18 This is where I got it from. I think we've been using "last updated" in the UI so going to keep it like this

**Pull request merged**